gists/frontend-pull-request-guidelines.md

```diff
@@ -0,0 +1,94 @@
+    ].filter(Boolean).join(" ")
+  }
+
+  formatName({
+    title: "Jr",
+    firstName: "Anthony",
+    lastName: "Soprano"
+  }) // Anthony Soprano Jr
+  ```
+- If there is a particularly complicated util / functionality, check to see if [lodash](https://lodash.com/docs/4.17.15) has something that fits your needs. If not, write unit tests
+- If something is no longer being used, it's better to delete it! The more code you have, the more bugs you have
+  - This applies to files, functions, lines of code, styles, and comments. If it turns out that you need it, you can retrieve it from the Git history
+- Avoid [variable shadowing](https://en.wikipedia.org/wiki/Variable_shadowing) (i.e., using the same name for variables in different scopes). This leads to code that can be hard to read and defects that are difficult to diagnose
+- Duplicating context or sowing confusion with a variable name (e.g., using a singular `participant` variable for an array of participants)
+  - Refer to the [variable naming cheatsheet](https://github.com/kettanaito/naming-cheatsheet) for more examples
+- You probably don't need [reduce](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce). Most applications of reduce can be simplified and more performant when using [local mutation](https://typeofnan.dev/mutation-isnt-always-bad-in-javascript/)
+
+## React
+- You should virtually **never** add an `onClick` prop to anything other than a `button` element
+  - The `button` element contains several accessibility features out-of-the-box, including keyboard support. Creating custom markup isn't worth it and requires a lot of additional attributes to reach parity with user's expectations
+- You probably don't need `useEffect` (unless you're fetching data). Setting state in a `useEffect` is typically code smell for a use case for deriving the state, instead of syncing it
+  - https://react.dev/learn/you-might-not-need-an-effect
+  - https://kentcdodds.com/blog/dont-sync-state-derive-it
+- Loose prop type definitions breed defects. Be as specific as you can when defining a prop type and place custom shapes in `types.js`
+  - You should virtually never use `PropTypes.object` or `PropTypes.array`, instead reach for `PropTypes.shape` and `PropTypes.arrayOf`, respectively
+- If a prop is _not_ required, it should have a sensible default declared in `defaultProps`
+- Only put truly global components in `src/js/components`. If a component only applies to one subsection, keep it in that subsection's `components` folder and only move it to the global component folder if another use case for it comes up
+- Try to avoid passing React state and state setters (e.g., `x`, `setX`) directly to another component
+  - This couples the child component to the parent's implementation and typically is more so just rearranging files, rather than aligning with the React component-paradigm by creating a reusable, composable interface
```

Comment:
Is there a good article for reference? Personally I find this to be one of the hardest things about React.

Reply:
Ugh I've been trying to find one forever. Maybe I should just write it